components/sup/src/manager/mod.rs

```diff
@@ -879,20 +904,30 @@ impl Manager {
         self.butterfly.restart_elections();
     }
 
-    fn shutdown(&self) {
+    fn shutdown(&mut self) {
         outputln!("Gracefully departing from butterfly network.");
         self.butterfly.set_departed();
 
-        let mut services = self.services.write().expect("Services lock is poisend!");
+        let mut svcs = Vec::new();
 
-        for mut service in services.drain(..) {
+        // The problem we're trying to work around here by adding this block is that `write`
+        // creates an immutable borrow on `self`, and `self.remove_service` needs `&mut self`.
+        // The solution is to introduce the block to drop the immutable borrow before the call to
+        // `self.remove_service`, and use `mem::swap` to move the services to a variable defined
+        // outside the block while we have the lock.
+        {
+            let mut services = self.services.write().expect("Services lock is poisoned!");
+            mem::swap(services.deref_mut(), &mut svcs);
+        }
```

Comment:
`remove_service` took `&self` before, rather than `&mut self`.

Now it takes `&mut self` because it calls `self.user_config_watcher.remove(service)`, which stops the watcher thread for that service and removes it from the `states` HashMap. Therefore, we have a clash between the above, immutable borrow, and the mutable one below. Hence the block.

Previou